File: src/calculator/expo/cards/effects/scarce-trail.ts
import type {
  DefaultCalculatorReturnType,
  GenericCalculatorParams,
} from '@nanosh/types/generic'
import GetRandomNumber from '@nanosh/utils/getRandomNumber'
import seedrandom from 'seedrandom'

interface ScarceTrailParams
  extends Pick<GenericCalculatorParams, 'state' | 'gameID' | 'invokeTime'> {}

const SCARCE_TRAIL_MIN_INTEL_BASIC = 0
const SCARCE_TRAIL_MAX_INTEL_BASIC = 1

export default function ScarceTrail({
  state,
  gameID,
  invokeTime,
}: ScarceTrailParams): DefaultCalculatorReturnType {
  const stateCopy = structuredClone(state)

  const prng = seedrandom(`${gameID}-${invokeTime}`)
  const intelBasicAdded = GetRandomNumber(
    SCARCE_TRAIL_MIN_INTEL_BASIC,
    SCARCE_TRAIL_MAX_INTEL_BASIC,
    prng,
  )

  stateCopy.intel.basic += intelBasicAdded

  return [stateCopy, null]
}

File: src/nakama/gamelogic/calculator/expedition/init/loot/scarce_trail.go
package loot

import (
	expeditionErrors "nanosh/nakama-modules/gamelogic/calculator/expedition"
	"nanosh/nakama-modules/gamelogic/types"
	"nanosh/nakama-modules/gamelogic/utils/dice"
	"nanosh/nakama-modules/gamelogic/utils/random"
)

// ScarceTrail - Gain 0-1 Basic Intel
func ScarceTrail(game *types.Game, gameID string, invokeTime int64, username string) (*types.Game, error) {
	// Check expedition exists
	if game.Expedition == nil {
		return nil, expeditionErrors.ErrNoActiveExpedition
	}

	// Create deterministic PRNG using standard pattern
	prng := random.CreateSeededPRNG(gameID, invokeTime, username)

	// Roll for intel gain (0-1)
	intelGain := dice.RollRange(prng, 0, 1)

	// Add Basic Intel
	game.Intel.Basic += intelGain

	return game, nil
}
